lib/Transforms/HandshakeHoistExtInstances.cpp

```diff
@@ -100,7 +110,14 @@ void HandshakeHoistExtInstancesPass::hoistInstances(handshake::FuncOp funcOp,
       continue;
 
     anyInstance = true;
-    StringRef instName = getUniqueName(instOp);
+    StringRef instFuncName = instFuncOp.getNameAttr().strref();
+
+    assert(std::find(externalFunctionNames.begin(), externalFunctionNames.end(),
+                     instFuncName) == externalFunctionNames.end() &&
+           "External function instantiated multiple times. Each external "
+           "function refers to one output and input port.");
+
+    externalFunctionNames.push_back(instFuncName);
```

Comment:
Use asserts only when testing for a piece of code's internal consistency (something that should really never fail if the code is correct), not for unsupported inputs (here an IR which contains multiple calls to the same external function). Asserts are removed in Release builds so in this case the pass would "successfully" produce an IR that we technically do not support. Change the method's return type to `LogicalResult` and produce a failure instead.

```cpp
// Using a set on operations for clearer semantics
SmallPtrSet<handshake::FuncOp, 4> calledFunctions;

// ...

if (auto [_, newFunc] : calledFunctions.insert(instFuncOp) ; newFunc) {
  return insertFuncOp.emitError() << "external function is instantiated"
   << "multiple times, but we only support a single instantation in any"
   << "given kernel";
}
```